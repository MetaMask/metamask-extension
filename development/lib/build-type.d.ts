import { Infer, Struct } from 'superstruct';

// TODO: Replace `any` with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Unique<Element extends Struct<any>> = (
  struct: Struct<Infer<Element>[], Infer<Element>>,
  eq?: (a: Infer<Element>, b: Infer<Element>) => boolean,
) => Struct<Infer<Element>[], Infer<Element>>;
