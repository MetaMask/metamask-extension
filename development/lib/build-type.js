const fs = require('fs');
const { AssertionError } = require('assert');
const path = require('path');
const {
  object,
  string,
  record,
  optional,
  array,
  refine,
  any,
  boolean,
  coerce,
  union,
  number,
  unknown,
  validate,
  nullable,
  never,
  literal,
} = require('superstruct');
const yaml = require('yaml');
const { uniqWith } = require('lodash');

const BUILDS_YML_PATH = path.resolve('./builds.yml');

/**
 * @type {import('superstruct').Infer<typeof BuildTypesStruct> | null}
 */
let cachedBuildTypes = null;

/**
 * Ensures that the array item contains only elements that are distinct from each other
 *
 * @template {Struct<any>} Element
 * @type {import('./build-type').Unique<Element>}
 */
const unique = (struct, eq) =>
  refine(struct, 'unique', (value) => {
    if (uniqWith(value, eq).length === value.length) {
      return true;
    }
    return 'Array contains duplicated values';
  });

const EnvDefinitionStruct = coerce(
  object({ key: string(), value: unknown() }),
  refine(record(string(), any()), 'Env variable declaration', (value) => {
    if (Object.keys(value).length !== 1) {
      return 'Declaration should have only one property, the name';
    }
    return true;
  }),
  (value) => ({ key: Object.keys(value)[0], value: Object.values(value)[0] }),
);

const EnvArrayStruct = unique(
  array(union([string(), EnvDefinitionStruct])),
  (a, b) => {
    const keyA = typeof a === 'string' ? a : a.key;
    const keyB = typeof b === 'string' ? b : b.key;
    return keyA === keyB;
  },
);

/**
 * Ensures a number is within a given range
 *
 * @param {number} min
 * @param {number} max
 */
const isInRange = (min, max) => {
  /**
   *
   * @param {number} value
   * @returns boolean
   */
  function check(value) {
    return value >= min && value <= max;
  }
  return refine(number(), 'range', check);
};

const BuildTypeStruct = object({
  id: isInRange(10, 64),
  features: optional(unique(array(string()))),
  env: optional(EnvArrayStruct),
  isPrerelease: optional(boolean()),
  manifestOverrides: union([string(), literal(false)]),
  buildNameOverride: union([string(), literal(false)]),
});

const CopyAssetStruct = object({ src: string(), dest: string() });
const ExclusiveIncludeAssetStruct = coerce(
  object({ exclusiveInclude: string() }),
  string(),
  (exclusiveInclude) => ({ exclusiveInclude }),
);
const AssetStruct = union([CopyAssetStruct, ExclusiveIncludeAssetStruct]);

const FeatureStruct = object({
  env: optional(EnvArrayStruct),
  // TODO(ritave): Check if the paths exist
  assets: optional(array(AssetStruct)),
});

const FeaturesStruct = refine(
  record(
    string(),
    coerce(FeatureStruct, nullable(never()), () => ({})),
  ),
  'feature definitions',
  function* (value) {
    let isValid = true;

    const definitions = new Set();

    for (const feature of Object.values(value)) {
      for (const env of feature?.env ?? []) {
        if (typeof env !== 'string') {
          if (definitions.has(env.key)) {
            isValid = false;
            yield `Multiple defined features have a definition of "${env}" env variable, resulting in a conflict`;
          }
          definitions.add(env.key);
        }
      }
    }
    return isValid;
  },
);

const BuildTypesStruct = refine(
  object({
    default: string(),
    buildTypes: record(string(), BuildTypeStruct),
    features: FeaturesStruct,
    env: EnvArrayStruct,
  }),
  'BuildTypes',
  (value) => {
    if (!Object.keys(value.buildTypes).includes(value.default)) {
      return `Default build type "${value.default}" does not exist in builds declarations`;
    }
    return true;
  },
);

/**
 * Loads definitions of build type and what they are composed of.
 *
 * @returns {import('superstruct').Infer<typeof BuildTypesStruct>}
 */
function loadBuildTypesConfig() {
  if (cachedBuildTypes !== null) {
    return cachedBuildTypes;
  }
  const buildsData = yaml.parse(fs.readFileSync(BUILDS_YML_PATH, 'utf8'));
  const [err, result] = validate(buildsData, BuildTypesStruct, {
    coerce: true,
  });
  if (err !== undefined) {
    throw new AssertionError({
      message: constructFailureMessage(err),
    });
  }
  cachedBuildTypes = result;
  return cachedBuildTypes;
}

/**
 * Creates a user readable error message about parse failure.
 *
 * @param {import('superstruct').StructError} structError
 * @returns {string}
 */
function constructFailureMessage(structError) {
  return `Failed to parse builds.yml
  -> ${structError
    .failures()
    .map(
      (failure) =>
        `${failure.message} (${BUILDS_YML_PATH}:.${failure.path.join('/')})`,
    )
    .join('\n  -> ')}
`;
}

module.exports = { loadBuildTypesConfig };
