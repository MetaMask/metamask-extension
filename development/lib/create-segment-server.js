const http = require('http');

/**
 * This is the default error handler to be used by this mock segment server.
 * It will print the error to the console and exit the process.
 *
 * @param {Error} error - The server error
 */
function defaultOnError(error) {
  console.log(error);
  process.exit(1);
}

/**
 * @typedef {import('http').IncomingMessage} IncomingMessage
 * @typedef {import('http').ServerResponse} ServerResponse
 */

/**
 * This function handles requests for the mock Segment server
 *
 * @typedef {(request: IncomingMessage, response: ServerResponse, metricEvents: Array<object>) => void} MockSegmentRequestHandler
 */

/**
 * Creates a HTTP server that acts as a fake version of the Segment API.
 * The bevahiour is rudimentary at the moment - it returns HTTP 200 in response
 * to every request. The only function this serves is to spy on requests sent to
 * this server, and to parse the request payloads as Segment batch events.
 *
 * @param {MockSegmentRequestHandler} onRequest - A callback for each request the server receives.
 * @param {(error: Error) => void} [onError] - A callback for server error events
 */
function createSegmentServer(onRequest, onError = defaultOnError) {
  const server = http.createServer(async (request, response) => {
    const chunks = [];

    request.on('data', (chunk) => {
      chunks.push(chunk);
    });

    await new Promise((resolve) => {
      request.on('end', () => {
        resolve();
      });
    });

    // respond to preflight request
    if (request.method === 'OPTIONS') {
      response.setHeader('Access-Control-Allow-Origin', '*');
      response.setHeader('Access-Control-Allow-Methods', '*');
      response.setHeader('Access-Control-Allow-Headers', '*');
      response.statusCode = 200;
      response.end();
      return;
    }

    let metricEvents = [];
    if (chunks.length) {
      const body = Buffer.concat(chunks).toString();
      const segmentPayload = JSON.parse(body);
      metricEvents = segmentPayload.batch;
    }

    onRequest(request, response, metricEvents);
  });

  server.on('error', onError);

  return {
    start: async (port) => {
      await new Promise((resolve, reject) => {
        server.listen(port, (error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
      });
    },
    stop: async () => {
      await new Promise((resolve, reject) => {
        server.close((error) => {
          if (error) {
            return reject(error);
          }
          return resolve();
        });
        // We need to close all connections to stop the server quickly
        // Otherwise it takes a few seconds for it to close
        server.closeAllConnections();
      });
    },
  };
}

module.exports = { createSegmentServer };
