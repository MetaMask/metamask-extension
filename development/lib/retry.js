/**
 * Re-runs the given function until it returns a resolved promise or the number
 * of retries is exceeded, whichever comes first (with an optional delay in
 * between retries).
 *
 * @param {object} args - A set of arguments and options.
 * @param {number} args.retries - The maximum number of times to re-run the
 * function on failure.
 * @param {number} [args.delay] - The amount of time (in milliseconds) to wait in
 * between retries. (Default: 0)
 * @param {string} [args.rejectionMessage] - The message for the rejected promise
 * this function will return in the event of failure. (Default: "Retry limit
 * reached")
 * @param {boolean} [args.stopAfterOneFailure] - Retries until the function fails.
 * @param {Function} functionToRetry - The function that is run and tested for
 * failure.
 * @returns {Promise<* | null | Error>} a promise that either resolves with one of the following:
 * - If successful, resolves with the return value of functionToRetry.
 * - If functionToRetry fails while stopAfterOneFailure is true, resolves with null.
 * - Otherwise it is rejected with rejectionMessage.
 */
async function retry(
  {
    retries,
    delay = 0,
    rejectionMessage = 'Retry limit reached',
    stopAfterOneFailure = false,
  },
  functionToRetry,
) {
  let attempts = 0;
  while (attempts <= retries) {
    if (attempts > 0 && delay > 0) {
      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    try {
      const result = await functionToRetry();
      if (!stopAfterOneFailure) {
        return result;
      }
    } catch (error) {
      if (error.message === "Exited with code '1'") {
        console.log("retry() received: Exited with code '1'");
      } else {
        console.error('error caught in retry():', error);
      }

      if (stopAfterOneFailure) {
        throw new Error('Test failed. No more retries will be performed');
      }

      if (attempts < retries) {
        console.log('Ready to retry() again');
      }
    } finally {
      attempts += 1;
    }
  }

  if (stopAfterOneFailure) {
    return null;
  }

  throw new Error(rejectionMessage);
}

module.exports = { retry };
