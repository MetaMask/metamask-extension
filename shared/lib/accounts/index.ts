export * from './accounts';
export * from './bitcoin-wallet-snap';
export * from './snaps';
export * from './solana-wallet-snap';
