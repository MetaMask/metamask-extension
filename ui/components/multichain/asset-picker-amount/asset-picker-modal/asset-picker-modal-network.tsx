import React from 'react';

import { useSelector } from 'react-redux';
import { NetworkConfiguration } from '@metamask/network-controller';
import {
  Display,
  FlexDirection,
  BlockSize,
} from '../../../../helpers/constants/design-system';
import {
  ModalOverlay,
  ModalContent,
  ModalHeader,
  Modal,
  Box,
} from '../../../component-library';
import { CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP } from '../../../../../shared/constants/network';
///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
import { useI18nContext } from '../../../../hooks/useI18nContext';
///: END:ONLY_INCLUDE_IF
import { NetworkListItem } from '../../network-list-item';
import { getNetworkConfigurationsByChainId } from '../../../../selectors';
import { getProviderConfig } from '../../../../ducks/metamask/metamask';

/**
 * AssetPickerModalNetwork component displays a modal for selecting a network in the asset picker.
 *
 * @param props
 * @param props.isOpen - Determines whether the modal is open or not.
 * @param props.network - The currently selected network, not necessarily the active wallet network.
 * @param props.networks - The list of selectable networks.
 * @param props.onNetworkChange - The callback function to handle network change.
 * @param props.onClose - The callback function to handle modal close.
 * @param props.onBack - The callback function to handle going back in the modal.
 * @returns A modal with a list of selectable networks.
 */
export const AssetPickerModalNetwork = ({
  isOpen,
  onClose,
  onBack,
  network,
  networks,
  onNetworkChange,
}: {
  isOpen: boolean;
  network?: NetworkConfiguration;
  networks?: NetworkConfiguration[];
  onNetworkChange: (network: NetworkConfiguration) => void;
  onClose: () => void;
  onBack: () => void;
}) => {
  ///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
  const t = useI18nContext();
  ///: END:ONLY_INCLUDE_IF

  const currentNetwork = useSelector(getProviderConfig);
  const allNetworks = useSelector(getNetworkConfigurationsByChainId);

  const selectedNetwork =
    network ?? (currentNetwork?.chainId && allNetworks[currentNetwork.chainId]);

  const networksList: NetworkConfiguration[] =
    networks ?? Object.values(allNetworks) ?? [];

  return (
    <Modal
      isOpen={isOpen}
      onClose={onClose}
      className="multichain-asset-picker__network-modal"
    >
      <ModalOverlay />
      <ModalContent modalDialogProps={{ padding: 0 }}>
        <ModalHeader onBack={onBack} onClose={onClose}>
          {t('bridgeSelectNetwork')}
        </ModalHeader>
        <Box className="multichain-asset-picker__network-list">
          <Box
            style={{
              gridColumnStart: 1,
              gridColumnEnd: 3,
            }}
            display={Display.Flex}
            flexDirection={FlexDirection.Column}
            height={BlockSize.Full}
          >
            {networksList.map((networkConfig) => {
              const { name, chainId } = networkConfig;
              return (
                <NetworkListItem
                  key={chainId}
                  name={name}
                  selected={selectedNetwork?.chainId === chainId}
                  onClick={() => {
                    onNetworkChange(networkConfig);
                    onBack();
                  }}
                  iconSrc={
                    CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP[
                      chainId as keyof typeof CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP
                    ]
                  }
                  focus={false}
                />
              );
            })}
          </Box>
        </Box>
      </ModalContent>
    </Modal>
  );
};
