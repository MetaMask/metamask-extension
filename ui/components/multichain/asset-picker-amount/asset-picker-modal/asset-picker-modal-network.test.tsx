import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { screen, fireEvent } from '@testing-library/react';
import { RpcEndpointType } from '@metamask/network-controller';
import { renderWithProvider } from '../../../../../test/lib/render-helpers';
import mockState from '../../../../../test/data/mock-send-state.json';
import { AssetPickerModalNetwork } from './asset-picker-modal-network';

const mockOnClose = jest.fn();
const mockOnNetworkChange = jest.fn();
const mockOnBack = jest.fn();

describe('AssetPickerModalNetwork', () => {
  const mockStore = configureStore([thunk]);
  const store = mockStore(mockState);

  const defaultProps = {
    isOpen: true,
    onClose: mockOnClose,
    onBack: mockOnBack,
    network: undefined,
    networks: [],
    onNetworkChange: mockOnNetworkChange,
  };

  const networkProps = {
    network: {
      chainId: '0x1',
      nativeCurrency: 'ETH',
      defaultBlockExplorerUrlIndex: 0,
      blockExplorerUrls: ['https://explorerurl'],
      defaultRpcEndpointIndex: 0,
      rpcEndpoints: [
        {
          networkClientId: 'test1',
          url: 'https://rpcurl',
          type: RpcEndpointType.Custom,
        },
      ],
      name: 'network',
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any,
    networks: [
      {
        chainId: '0x1',
        nativeCurrency: 'ETH',
        defaultBlockExplorerUrlIndex: 0,
        blockExplorerUrls: ['https://explorerurl'],
        defaultRpcEndpointIndex: 0,
        rpcEndpoints: [
          {
            networkClientId: 'test1',
            url: 'https://rpcurl',
            type: RpcEndpointType.Custom,
          },
        ],
        name: 'Network name 3',
      },
      {
        chainId: '0xa',
        nativeCurrency: 'ETH',
        defaultBlockExplorerUrlIndex: 0,
        blockExplorerUrls: ['https://explorerurl'],
        defaultRpcEndpointIndex: 0,
        rpcEndpoints: [
          {
            networkClientId: 'test2',
            url: 'https://rpcurl',
            type: RpcEndpointType.Custom,
          },
        ],
        name: 'Network name 4',
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ] as any,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders modal with no network list by default', () => {
    const { baseElement } = renderWithProvider(
      <AssetPickerModalNetwork {...defaultProps} />,
      store,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should not show selected network when network prop is not passed in', () => {
    const { baseElement } = renderWithProvider(
      <AssetPickerModalNetwork
        {...defaultProps}
        networks={networkProps.networks}
      />,
      store,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should use passed in network as default when network prop is passed in', () => {
    const { baseElement } = renderWithProvider(
      <AssetPickerModalNetwork {...defaultProps} {...networkProps} />,
      store,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should call onClose and onBack when header buttons are clicked', () => {
    renderWithProvider(<AssetPickerModalNetwork {...defaultProps} />, store);

    fireEvent.click(screen.getByLabelText('Close'));
    expect(mockOnClose).toHaveBeenCalledTimes(1);

    fireEvent.click(screen.getByLabelText('Back'));
    expect(mockOnBack).toHaveBeenCalledTimes(1);
  });

  it('should call onBack and onClickHandler when network is selected', () => {
    renderWithProvider(
      <AssetPickerModalNetwork {...defaultProps} {...networkProps} />,
      store,
    );

    fireEvent.click(screen.getByText('Network name 3'));
    expect(mockOnBack).toHaveBeenCalledTimes(1);
    expect(mockOnNetworkChange).toHaveBeenCalledTimes(1);
  });
});
