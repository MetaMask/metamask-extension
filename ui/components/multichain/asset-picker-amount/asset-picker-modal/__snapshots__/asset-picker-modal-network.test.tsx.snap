// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetPickerModalNetwork renders modal with no network list by default 1`] = `
<body>
  <div
    id="popover-content"
  />
  <div />
  <div
    class="mm-modal multichain-asset-picker__network-modal"
  >
    <div
      aria-hidden="true"
      class="mm-box mm-modal-overlay mm-box--width-full mm-box--height-full mm-box--background-color-overlay-default"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="mm-box mm-modal-content mm-box--padding-top-4 mm-box--sm:padding-top-8 mm-box--md:padding-top-12 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--sm:padding-bottom-8 mm-box--md:padding-bottom-12 mm-box--padding-left-4 mm-box--display-flex mm-box--justify-content-center mm-box--align-items-flex-start mm-box--width-screen mm-box--height-screen"
      >
        <section
          aria-modal="true"
          class="mm-box mm-modal-content__dialog mm-modal-content__dialog--size-sm mm-box--padding-0 mm-box--padding-top-4 mm-box--padding-bottom-4 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--background-color-background-default mm-box--rounded-lg"
          role="dialog"
        >
          <header
            class="mm-box mm-header-base mm-modal-header mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--justify-content-space-between"
          >
            <div
              class="mm-box"
              style="min-width: 0px;"
            >
              <button
                aria-label="Back"
                class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
              >
                <span
                  class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
                  style="mask-image: url('./images/icons/arrow-left.svg');"
                />
              </button>
            </div>
            <div
              class="mm-box mm-box--width-full"
            >
              <h4
                class="mm-box mm-text mm-text--heading-sm mm-text--text-align-center mm-box--color-text-default"
              >
                Select network
              </h4>
            </div>
            <div
              class="mm-box mm-box--display-flex mm-box--justify-content-flex-end"
              style="min-width: 0px;"
            >
              <button
                aria-label="Close"
                class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
              >
                <span
                  class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
                  style="mask-image: url('./images/icons/close.svg');"
                />
              </button>
            </div>
          </header>
          <div
            class="mm-box multichain-asset-picker__network-list"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--height-full"
              style="grid-column-start: 1; grid-column-end: 3;"
            />
          </div>
        </section>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`AssetPickerModalNetwork should not show selected network when network prop is not passed in 1`] = `
<body>
  <div
    id="popover-content"
  />
  <div />
  <div
    class="mm-modal multichain-asset-picker__network-modal"
  >
    <div
      aria-hidden="true"
      class="mm-box mm-modal-overlay mm-box--width-full mm-box--height-full mm-box--background-color-overlay-default"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="mm-box mm-modal-content mm-box--padding-top-4 mm-box--sm:padding-top-8 mm-box--md:padding-top-12 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--sm:padding-bottom-8 mm-box--md:padding-bottom-12 mm-box--padding-left-4 mm-box--display-flex mm-box--justify-content-center mm-box--align-items-flex-start mm-box--width-screen mm-box--height-screen"
      >
        <section
          aria-modal="true"
          class="mm-box mm-modal-content__dialog mm-modal-content__dialog--size-sm mm-box--padding-0 mm-box--padding-top-4 mm-box--padding-bottom-4 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--background-color-background-default mm-box--rounded-lg"
          role="dialog"
        >
          <header
            class="mm-box mm-header-base mm-modal-header mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--justify-content-space-between"
          >
            <div
              class="mm-box"
              style="min-width: 0px;"
            >
              <button
                aria-label="Back"
                class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
              >
                <span
                  class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
                  style="mask-image: url('./images/icons/arrow-left.svg');"
                />
              </button>
            </div>
            <div
              class="mm-box mm-box--width-full"
            >
              <h4
                class="mm-box mm-text mm-text--heading-sm mm-text--text-align-center mm-box--color-text-default"
              >
                Select network
              </h4>
            </div>
            <div
              class="mm-box mm-box--display-flex mm-box--justify-content-flex-end"
              style="min-width: 0px;"
            >
              <button
                aria-label="Close"
                class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
              >
                <span
                  class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
                  style="mask-image: url('./images/icons/close.svg');"
                />
              </button>
            </div>
          </header>
          <div
            class="mm-box multichain-asset-picker__network-list"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--height-full"
              style="grid-column-start: 1; grid-column-end: 3;"
            >
              <div
                class="mm-box multichain-network-list-item mm-box--padding-top-4 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--gap-4 mm-box--justify-content-space-between mm-box--align-items-center mm-box--width-full mm-box--background-color-transparent"
              >
                <div
                  class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-network mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-background-default box--border-style-solid box--border-width-1"
                >
                  <img
                    alt="Network name 3 logo"
                    class="mm-avatar-network__network-image"
                    src="./images/eth_logo.svg"
                  />
                </div>
                <div
                  class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--justify-content-flex-start mm-box--align-items-flex-start mm-box--width-full"
                  style="overflow: hidden;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--align-items-center mm-box--width-full"
                    data-testid="Network name 3"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md mm-text--ellipsis mm-box--color-text-default mm-box--background-color-transparent"
                      tabindex="0"
                    >
                      Network name 3
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="mm-box multichain-network-list-item mm-box--padding-top-4 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--gap-4 mm-box--justify-content-space-between mm-box--align-items-center mm-box--width-full mm-box--background-color-transparent"
              >
                <div
                  class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-network mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-background-default box--border-style-solid box--border-width-1"
                >
                  <img
                    alt="Network name 4 logo"
                    class="mm-avatar-network__network-image"
                    src="./images/optimism.svg"
                  />
                </div>
                <div
                  class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--justify-content-flex-start mm-box--align-items-flex-start mm-box--width-full"
                  style="overflow: hidden;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--align-items-center mm-box--width-full"
                    data-testid="Network name 4"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md mm-text--ellipsis mm-box--color-text-default mm-box--background-color-transparent"
                      tabindex="0"
                    >
                      Network name 4
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`AssetPickerModalNetwork should use passed in network as default when network prop is passed in 1`] = `
<body>
  <div
    id="popover-content"
  />
  <div />
  <div
    class="mm-modal multichain-asset-picker__network-modal"
  >
    <div
      aria-hidden="true"
      class="mm-box mm-modal-overlay mm-box--width-full mm-box--height-full mm-box--background-color-overlay-default"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="mm-box mm-modal-content mm-box--padding-top-4 mm-box--sm:padding-top-8 mm-box--md:padding-top-12 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--sm:padding-bottom-8 mm-box--md:padding-bottom-12 mm-box--padding-left-4 mm-box--display-flex mm-box--justify-content-center mm-box--align-items-flex-start mm-box--width-screen mm-box--height-screen"
      >
        <section
          aria-modal="true"
          class="mm-box mm-modal-content__dialog mm-modal-content__dialog--size-sm mm-box--padding-0 mm-box--padding-top-4 mm-box--padding-bottom-4 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--background-color-background-default mm-box--rounded-lg"
          role="dialog"
        >
          <header
            class="mm-box mm-header-base mm-modal-header mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--justify-content-space-between"
          >
            <div
              class="mm-box"
              style="min-width: 0px;"
            >
              <button
                aria-label="Back"
                class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
              >
                <span
                  class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
                  style="mask-image: url('./images/icons/arrow-left.svg');"
                />
              </button>
            </div>
            <div
              class="mm-box mm-box--width-full"
            >
              <h4
                class="mm-box mm-text mm-text--heading-sm mm-text--text-align-center mm-box--color-text-default"
              >
                Select network
              </h4>
            </div>
            <div
              class="mm-box mm-box--display-flex mm-box--justify-content-flex-end"
              style="min-width: 0px;"
            >
              <button
                aria-label="Close"
                class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
              >
                <span
                  class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
                  style="mask-image: url('./images/icons/close.svg');"
                />
              </button>
            </div>
          </header>
          <div
            class="mm-box multichain-asset-picker__network-list"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--height-full"
              style="grid-column-start: 1; grid-column-end: 3;"
            >
              <div
                class="mm-box multichain-network-list-item multichain-network-list-item--selected mm-box--padding-top-4 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--gap-4 mm-box--justify-content-space-between mm-box--align-items-center mm-box--width-full mm-box--background-color-primary-muted"
              >
                <div
                  class="mm-box multichain-network-list-item__selected-indicator mm-box--background-color-primary-default mm-box--rounded-pill"
                />
                <div
                  class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-network mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-background-default box--border-style-solid box--border-width-1"
                >
                  <img
                    alt="Network name 3 logo"
                    class="mm-avatar-network__network-image"
                    src="./images/eth_logo.svg"
                  />
                </div>
                <div
                  class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--justify-content-flex-start mm-box--align-items-flex-start mm-box--width-full"
                  style="overflow: hidden;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--align-items-center mm-box--width-full"
                    data-testid="Network name 3"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md mm-text--ellipsis mm-box--color-text-default mm-box--background-color-transparent"
                      tabindex="0"
                    >
                      Network name 3
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="mm-box multichain-network-list-item mm-box--padding-top-4 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--gap-4 mm-box--justify-content-space-between mm-box--align-items-center mm-box--width-full mm-box--background-color-transparent"
              >
                <div
                  class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-network mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-background-default box--border-style-solid box--border-width-1"
                >
                  <img
                    alt="Network name 4 logo"
                    class="mm-avatar-network__network-image"
                    src="./images/optimism.svg"
                  />
                </div>
                <div
                  class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--justify-content-flex-start mm-box--align-items-flex-start mm-box--width-full"
                  style="overflow: hidden;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--align-items-center mm-box--width-full"
                    data-testid="Network name 4"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md mm-text--ellipsis mm-box--color-text-default mm-box--background-color-transparent"
                      tabindex="0"
                    >
                      Network name 4
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</body>
`;
